akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala

```diff
@@ -0,0 +1,61 @@
+package akka.persistence.typed.internal
+
+import akka.actor.typed.Behavior
+import akka.actor.{ DeadLetter, StashOverflowException }
+import akka.actor.typed.scaladsl.{ ActorContext, MutableStashBuffer }
+import akka.event.{ Logging, LoggingAdapter }
+import akka.persistence._
+import akka.persistence.typed.internal.EventsourcedBehavior.UnstashedCommand
+import akka.{ actor ⇒ a }
+
+trait EventsourcedStashManagement {
+  import akka.actor.typed.scaladsl.adapter._
+
+  protected def log: LoggingAdapter
+
+  def extension: Persistence
+  def context: ActorContext[Any]
+
+  protected val internalStash: MutableStashBuffer[Any]
+
+  /**
+   * The returned [[StashOverflowStrategy]] object determines how to handle the message failed to stash
+   * when the internal Stash capacity exceeded.
+   */
+  protected def internalStashOverflowStrategy: StashOverflowStrategy =
+    extension.defaultInternalStashOverflowStrategy // FIXME don't support this
+
+  protected def stash(msg: Any, behavior: Behavior[Any]): Behavior[Any] = {
+    try internalStash.stash(msg) catch {
+      case e: StashOverflowException ⇒
+        internalStashOverflowStrategy match {
+          case DiscardToDeadLetterStrategy ⇒
+            // val snd = sender() // FIXME there is no sender
+            val snd: a.ActorRef = null // FIXME
+            context.system.deadLetters.tell(DeadLetter(msg, snd, context.self.toUntyped), snd) // FIXME I guess that's wrong
+
+          case ReplyToStrategy(response) ⇒
+            // sender() ! response // FIXME there is no sender
+            throw new RuntimeException("ReplyToStrategy does not make sense at all in Akka Typed, since there is no sender()!")
+          case ThrowOverflowExceptionStrategy ⇒
+            throw e
+        }
+    }
+
+    behavior
+  }
+
+  protected def unstash(ctx: ActorContext[Any], behavior: Behavior[Any]): Behavior[Any] = {
+    log.info(s"UN-STASH: ${internalStash.head} from (before unstash)" + internalStash)
+    if (internalStash.nonEmpty) internalStash.unstash(context, behavior, 1, UnstashedCommand(_))
```

Comment:
you can start by using unstashAll and then no wrapping is needed, and then we can look into optimizing by partial unstashing